### ISSUE TITLE
Get rid of dependency on Symbolic toolbox

### ISSUE THREAD

**Issue by pgagarinov** [SystemAnalysisDpt-CMC-MSU/ellipsoids#29]
Deadline is 3rd of November
--------------------------------------------------------------------------------------------------------
Symbolic toolbox is currently used by Ellipsoidal Toolbox for a variable replacement when we need to transform
an equation \dot{x}=A(t)x(t)+... for a solvability problem into an equation for a rachability problem. This is done in
\products\+elltool\+reach\ReachContinuous.m	216	t = sym('t');

Also there is a function 
\products\+gras\+mat\+fcnlib\iscellofstringconst.m

that is used for checking if matrix function doesn't depent on time i.e. constant.

The goal is to get rid of dependency on symbolic math by writing a new function

gras.sym.varreplace that does a replacement in an expression of a specified named variable with another expression.
Examples:
sin(t): t -> 10-t: sin(10-t)
[(sin(t)+t)/tt]^att: t -> 10-t: [(sin(10-t)+10-t)/tt]^att
sin(s): s->s/10 : sin(s/10)

These are just a few examples to show that it is not as trivial as replacing one symbol. You should think of more tricky cases.
This function should be coreved with tests (that test this function on each tricky example) placed into gras.sym.test.mlunit.BasicTC test case.

Once this function is implemented please implement one more function: gras.sym.isdepdendent which does the same as iscellofstringconst from
above but via gras.sym.varreplace this time. gras.sym.isdepdendent also needs to be covered with tests in the same test case.

Finally remove iscellofstringconst and replace all calls to iscellofstringconst with calls to isdepdendent. Also please refactor code that
uses sym function from Matlab toolbox with varreplace function. Then re-install Matlab without Symbolic toolbox and make sure that 
all tests still pass after your changes.

**Comment by Vault-Guy** [comment]
There are several questions аbout the problem.
First of all, please, explain me the second example. Do you mean [(sin(t)+t)/tt]^ttt -> [(sin(10-t)+10-t)/(10-t)^2]^(10-t)^3 ?
Second one: what is the doubt of symbolic variable "t", used on the strings 216 and 217?
Perhaps your answers will help me to understand the rest part so I stop.

**Comment by pgagarinov** [comment]
1)>First of all, please, explain me the second example.
There was a misprint in the description. The idea is that "att" should not be replaced".

2) Now sure I understand the question, what do you mean by "doubt"?

You mean why matlab editor complains? This is because "eval" is used to replace the variable indirectly and there is no any direct reference to "t". 

If you ask why we need to get rid of dependency on Symbolic toolbox in the first place - just because it is an overkill to use the whole toolbox just to replace t with 10-t.

**Comment by Vault-Guy** [comment]
1) Oh, I understood! You meant that "t", "tt" and "att" are the different variables. OK, all right.

2) I don't understand, how and why we use this "t". We initialize it, but after that there is no using of it, as i see.

**Comment by pgagarinov** [comment]
2)  Here is an example to help you understand (in our case formulaStr is formed externally and passed as a parameter).

>> t=1

t =

     1

>> formulaStr='t+1';
>> eval(formulaStr)

ans =

     2

**Comment by Vault-Guy** [comment]
No, thank you, but I know what does eval do. 
Oooooh. It's fantastic Am I right that in the string 219 evCMat = cellfun(@eval, strCMat, 'UniformOutput', false); we replace all the entries of "t" in strCMat with "10-t" (if tSum = 10), yeah?

**Comment by pgagarinov** [comment]
Yeap)

**Comment by Vault-Guy** [comment]
Several new questions.
1) Should I write any new tests for testing of functions? 
2) There is no alternative for reinstalling of Matlab without Symbolic Toolbox? It's very surprisingly if there is no possibility just to switch off toolboxes.
3) The date of deadline appeared in the day of deadline, as I see.. Is is fair?) In any situation, please, give me the time till the Monday, I had planned to finish the task on the weekend!

**Comment by pgagarinov** [comment]
1) Yes, every new function needs to be covered with tests
2) You can install 32bit version without Symbolic toolbox and use it along with 64bit version with the toolbox.
3) It is my fault that I forgot to specify the deadline for your task in the very beginning. But the deadline for your task is later than most of other students have and the task is pretty simple. But since I specified the deadline too late let us push it forward by a few days and make it Tuesday. I think this will be fair given simplicity of the task.

**Comment by pgagarinov** [comment]
Btw - I think the best way to do the variable replacement is via regular expressions (Matlab regexp function).

**Comment by Vault-Guy** [comment]
What if I don't need a parameter "isDiscrete" in the function "iscellofstringconst"?

**Comment by pgagarinov** [comment]
What do you mean?

**Comment by Vault-Guy** [comment]
As I understand, we used parameter isDiscrete just to determine what a symbolic variable can be found in mCMat. But via using of regular expression we can find at the same time entries of 'k' and of 't' and understand if mCMat dependent. Am i right?

**Comment by pgagarinov** [comment]
Well, nothing prevented authors of iscellofstringconst to the same and make the function replace both k and t. However it is not 100% robust - if you use "t" for a discrete system by mistake your function won't throw an exception while iscellofstringconst will. So I think we better keep the interface of the function intact and just change the body.

**Comment by Vault-Guy** [comment]
OK, now it's clear, thank you.

**Comment by Vault-Guy** [comment]
As I see, there is no folder "sym" in "gras". I should create it, yes?

**Comment by pgagarinov** [comment]
Yes, please.

**Comment by Vault-Guy** [comment]
And again I have several questions.
1) Can I be sure that i won't meet matrix named "t" in getBackwardStrCMAt?
2) About regular expressions. How can I rescan any portions of thw string? For example:

str  = '1+t+t^2';
myExp = '\Wt\W';
regexp (str, myExp)

ans = 2

i.e. @regexp doesn't match '+t^', but I need it. Is it possible? I can't find any useful flags, modes or something else. 
Of course, it can be done through writing of the string parser, but I don't think that it's a good idea.

**Comment by pgagarinov** [comment]
1) This method returns C(t) which is supposed to contain only numbers and time variable. So I think the answer is "yes, you can" by design.

2) if you do not specify 'once' to regexp it finds all entries of "t", not just the first one.. \W is not supposed to match ^ so it is only natural that you cannot identify the second "t". 
Here is what Matlab doc says about \W "Any character that is not alphabetic, numeric, or underscore. For English character sets, \W is equivalent to [^a-zA-Z_0-9]".

 I think you need to read more about regular expressions in Matlab doc and figure out by yourself how to build a regular expression that works in this case. I could do it for you but this is not right - you should do it by yourself, this is your task after all.

**Comment by Vault-Guy** [comment]
1) Thank you!
2.1) \W matches symbol '^', I have already tested that;
2.2) My question is about rescanning of string parts, I don't ask you to write the who;e regular expression, I'm not so impudent) But:

str = '1+t+t*2';
myExp = '\Wt\W';
regexp (str, myExp)
ans = 2

I want to say that regexp doesn't match substrings with intersections like '+t+' and '+t*. The question is simple: is there any modes/flags/s.e. in regexp that can make regexp able to find all looking-for substrings despite of intersections? I just tired of reading help, and I want to understand is it possible or not, I don't ask for whole regulae expression)

**Comment by pgagarinov** [comment]
A small hint - see "Quantifiers" section, greedy expression mode.

**Comment by Vault-Guy** [comment]
Is it all?)

**Comment by pgagarinov** [comment]
Not quite)

3) in varreplace please add a brief description for all parameters, please do not use spaces

isstrprop(tempStr (j - 1), 'alphanum') && ...

I suspect that someone taught you to use a lot of spaces but in ET please don't. The convention is not to use spaces everywhere.
So please remove spaces from everywhere you added them.

4) in varreplace i, j are bad names for iterators - read Wiki

5) in varreplace 

mCMat(i) = strcat('0+', mCMat(i), '+0'); is inefficient and a bad style, please use 

mCMat{i} = ['0+', mCMat{i}, '+0'];

6) Please apply automatic indentation (Ctrl+I, remember?) to all the code you wrote.
 
7) In varreplace tempStr = char(mCMat(i)); is ugly, use tempStr = mCMat{i}; 

8) once you fix 5) and 7) you won't need reverseCMat = cell(tempCMat); in varreplace. 

9) in varreplace (and everywhere) please tend to use numel instead of length

10) in varreplace please also use regular expressions. I guess you need to implement 
a generic function "findvarind" that that finds indices of specified variable entries in a string.
Once you call it like this indVec=findvarind(..) you can then use strrep to replace "t" with "10-t".
The you can use this function in both varreplace and isdependent. isstrprop should not be used.

Or you can use regexprep to replace strings via regular expression directly. 


A small hint: use grouping operator "|" and beginning of word and end of word ancors. If you form your regexp correctly 
it will work like this:

>> regexp('t+t+tt+t','secret regular expression')

ans =

     1     3     7

11) Many of 3-10) apply to isdependent function as well.

**Comment by Vault-Guy** [comment]
Ok, i succeeded in getting of vector of indices of variable 't' in each string by using of 'secret regular expression' - thank you for advice about end-begin word anchors. But now new question appeared: you told me about using of strrep after that but strrep doesn't accept vector of indices as input argument. Am i really need vecInd?

**Comment by Vault-Guy** [comment]
And is it possible to use function 'break' to speed up work of developing function?

**Comment by pgagarinov** [comment]
12) Let us just use regexprep:

 
>> regexprep('t+t+t','t','b')

ans =

b+b+b

13) Yes, especially if it is in a loop.

**Comment by Vault-Guy** [comment]
Ha, i found a problem. I could finished this task more than week later. But I didn't know that 't+t' consist of 2 words 't' and plus, I thought we suppose it to be one word.
But now I have totally finished my task, I hope.

**Comment by pgagarinov** [comment]
This is much better, a few last fixes to be made:

14) varreplace - output name in the help header doesn't match the real variable name
15) repStr=['(',num2str(tSum),'-t)']; - > sprintf('(%d)',tSum)
16) for iElem=1:nElems - > cellfun(...,'UniformOutput',false)
3) is not done
17) Please use tabs instead of spaces for indentation - see my commit to your banch. (Pull with "launch rebase after fetch" flag)
18) There should be a single space followed by tabs (0 or more)
19) No help header for isdependent
20) Please do not use ";" after end

**Comment by Vault-Guy** [comment]
Please, just tell me that I have finished this task)

**Comment by pgagarinov** [comment]
21) you messed up the help header for varreplace, why?
22) again, in the help header for isdepended you used spaces, not tabs for indentation - see my changes to varreplace help header for a correct formatting.
23) @(X) regexprep(X,regExpression,repStr) -> @(x) regexprep(x,regExpression,repStr)
24) You have a lot of test faileres in your branch

**Comment by pgagarinov** [comment]
25) Also, you didn't do rebase for a long time, your branch is behind master by 15 commits

**Comment by Vault-Guy** [comment]
21) I missed your commit, I tried to unite your commit with mine in one. Perhaps, i've done something wrong.
22) Oh, I understood. I didn't unmark the 'Tab key inserts spaces'. I coudn't understand what was unright.
23) Fixed.
24) I try to do my best in debug.
25) I will rebase after debug.

**Comment by Vault-Guy** [comment]
And now?

**Comment by pgagarinov** [comment]
What about re-base?

**Comment by Vault-Guy** [comment]
What will happen after push if I have such log? https://yadi.sk/i/ECZvP7HkkR6SU

**Comment by pgagarinov** [comment]
Nothing good...

I suspect that you forgot to check "Launch rebase after fetch" checkbox when doing pull. You should not use sync or just pull with merge. Always use pull with rebase when you need to get the latest changes from origin/yourbranch into origin/yourbranch.

Here is how you fix this:

1) Select the commit that you made on 11.11.2015 at 16.04 (second from the top), right click and select "reset to this" with "hard mode"
2) Do pull with "Launch rebase after fetch".
4) Rebase your local branch to master and push

**Comment by Vault-Guy** [comment]
Now I have this: https://yadi.sk/i/dSaocn9SkRCF4

But also:
$ git status
On branch Issue_29_NikitaL
Your branch and 'origin/Issue_29_NikitaL' have diverged,
and have 35 and 16 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)
nothing to commit, working directory clean

But then I pull nothing happens. What can I do?

**Comment by pgagarinov** [comment]
Ok, 
1) the structure is good so no worries.

2) if you check "rebase after fetch" on your "pull" dialog and then go through rebase you should be able to get the latest changes from origin/Issue_29_NikitaL into your branch. There is no any other way around. if it doesn't help then it just means you do something wrong. The key part here is rebase - if it runs and successful - you should end-up with your branch being newer than origin's. After that you should push.

**Comment by Vault-Guy** [comment]
I'm sure, this version is ready for master branch.

**Comment by pgagarinov** [comment]
Not even close, I don't know what you did but now your branch looks like a mess. Your branch need to look like this after rebase: 

your commit
your commit
your commit
---------------------
not your commit
not your commit
not your commit

Moreover the part that is below ----------- should look exactly like master branch. The whole Idea of Rebase is to make your branch look like master+ delta where delta consists of your commits only. 

Please note - you ran into these problems only because you did use merge. Please do not use merge until you need to merge you changes back to master. Pull does merge, sync can do merge. My recommendation for later:

a) Do not user sync, use pull with "launch rebase after fetch flag"
b) do not use merge until your task is ready for master.

Now here is what you need to do to fix things:

Approach A)

1) Switch to your branch
2) Open TortoiseGit-> Show Reglog
3) Find the latest commit that corresponded to a correct state of your branch
4) Right click - >switch/checkout to this. This will create a branch with the latest correct state. Then you can add the changes you might have added after (if any). Then rebase this branch to master and push it into issue_29_NikitaL

Approach B)

1) create a new branch named issue_29_NikitaL_2 from master.
2) switch to this branch
3) right click ->TortoiseGit->Browse references
4) Find remotes/origin/issue_29_NikitaL and open log for this branch
5) Select only your commits by keeping "Ctrl" pressed (multi-line selection). Do not select commits made by me or Ilya, only your commits. 
6) Right click - cherry pick
7) Go through a series of conflict resolutions

After the end of this painful procedure you will have your branch issue_29_NikitaL_2 that will look like it should.

**Comment by Vault-Guy** [comment]
Ooops, I did it again, I did something wrong(
I'm ready for all painful procedures of thw world, bot now I'm in the state of cherry-picking (I don't know why and how) and don't know how to exit. "git cherry-pick -quit" and "git reset --merge" didn't help (I haven't done "merge", but you never know waht happened). What can I do?

**Comment by pgagarinov** [comment]
git cherry-pick --abort

**Comment by Vault-Guy** [comment]
https://yadi.sk/i/21ExGi5ckSfB7

**Comment by pgagarinov** [comment]
try git reset HEAD --hard
if it doesn't help - do a fresh checkout

**Comment by Vault-Guy** [comment]
OK, I'm afraid of doing new fantastic mistakes, so new foolish question: then I rebase and conflicts appeared, is it normal to solve it through using TortoiseGitMerge?

**Comment by pgagarinov** [comment]
if you push to the branch that is not just a tail of your branch you need to specify "force ignore known chagnes". 

It doesn't matter which tool you use to resolve conflicts as long as you do rebase. I think you should avoid  using the command line interface for now and use TortoiseGIT. When you do rebase via TortoiseGIT you will see certain files marked red (in case of conflicts). If you click on those files a special tool for conflict resolution appears. Use this tool.

**Comment by Vault-Guy** [comment]
You mean: "Force: may discard known changes", yes?
OK, i have been using Tortoise for this deals from the beginning. I mean if it is normal to merge at all.

**Comment by pgagarinov** [comment]
Rebase and merge are different things, you use rebase to get changes from master and you use merge to put your changes back to master. That's it.

**Comment by Vault-Guy** [comment]
Yes, I know it. Perhaps I just can't formulate my thoughts.

But I still can't push though I've marked "force: may discard known changes". The same fault.

**Comment by pgagarinov** [comment]
You need to fetch first to make the changes known.

**Comment by Vault-Guy** [comment]
Fetch didn't help me. I'm still can't push "Issue_29_NikitaL" branch

**Comment by pgagarinov** [comment]
I'm a bit dissapointed by your inability to do elementary stuff so I decided to fix the problem by myself, it took 3 minutes. I pushed the fixed branch into "Issue_29_NikitaL". Now do fetch and check out this branch. This is the last time I do this.

Please do not screw this up again)

I expect you on next Tuesday with a laptop so that we can go through our git workflow step by step.

**Comment by Vault-Guy** [comment]
Mmmm.. Issue is closed, yes?

**Comment by pgagarinov** [comment]
Not yet, because you there is a lot of test failures in your branch - you need to fix them. i.e. 24) from above is not done. Then you need to make sure that all tests pass. Then you need to rebase again.

And ffter that you need to merge your branch into master following these steps:

1) Switch to master
2) Merge -> from -> origin/remotes/your branch (using local branch can dangerous because it can      contain unpushed changes) 
3) Important !!! Set "no fast forward flag". Press OK
4) DO NOT PUSH JUST YET. Right click and open log. It should look like "petlya".
5) If eveything is ok - press push. 
6) Wait for the test results for updated master. If everything ok - switch to the next task which will be assigned to your very shortly.

**Comment by pgagarinov** [comment]
I reverted your changes to master. You cannot merge to master until all tests in your branch pass.
Right now they do not pass so you broke master. Please do not do that again. 

STEP1: FIX BUGS IN YOUR BRANCH

STEP2: WAIT FOR TESTS RESULTS FOR YOUR BRANCH

STEP3: IF ALL TESTS PASS  - GO TO STEP 4. IF NOT - GO TO STEP1.

STEP4. MERGE TO MASTER.

**Comment by pgagarinov** [comment]
Well, master's part of your branch matches master and you have delta on top of it so this is good. But your delta looks incorrect because it is not a straight line. This is probably because you either used pull without "launch rebase after fetch" flag or just used merge instead of rebase at some stage. Please note that. Your delta (your commits) should form a straight line, if I do something in your branch and you need to pull this - please use "launch rebase after fetch" flag in pull dialog.

**Comment by Vault-Guy** [comment]
Well, I have a problem. There is some strange problem: even if current folder is '+fcnlib' I can't launch in Command window 'isdependent(c)'  - I see "Undefined function of variable 'c' ". And also some failures in gras.mat.test.run_tests appeared (such as you can see here: https://yadi.sk/i/NpQoKKffkcERq).
As I think that's a last my problem. I very, very want to finish the issue.

**Comment by pgagarinov** [comment]
You need to figure out what is wrong in your branch all by yourself - all tests is master pass so these falures are purely a result of your changes.

The easiest way is to use Reflog as I described above. 

You need to finish this issue before starting working on next one.

**Comment by Vault-Guy** [comment]
You can say nothing.

**Comment by pgagarinov** [comment]
Sure, you can delete your remove branch

**Comment by Vault-Guy** [comment]
Well, thank you!

**Comment by pgagarinov** [comment]
You committed directly to master which is prohibited by our project policy. I revoked your write access till Tuesday (inclusive) as a penalty. Next time, If you continue violating our Git commit policy, I'll revoke your write access for 1 week. I suggest you spend this time for learning Git, TorgotiseGit and our issue workflow (https://github.com/SystemAnalysisDpt-CMC-MSU/ellipsoids/wiki/Issue-workflow-policy-with-GIT). 

Please note that I'd like to avoid using penalties like this but given that this is the fifth time I have to revert someone's changes from master there is no really other choice. I hope you understand.

**Comment by Vault-Guy** [comment]
I have already updated master, rebased my branch, tested work without Symbolic Toolbox and begun to wait end of handling of my commits by server. Am I right that I can merge into master if succeed?

**Comment by pgagarinov** [comment]
Help headers for both isdependent and varreplace are not 100% correct (see below).

Bad:

% Input:
%	regular:
%			mCMat: cell[nDims1,nDims2] - 
%				cell matrix of system
%			isDiscrete: logical[1,1] - logical variable
%				which equals 1 in case of discrete system
%				and equals 0 in case of continious system


Good:

% Input:
%	regular:
%		mCMat: cell[nDims1,nDims2] - 
%			cell matrix of system
%		isDiscrete: logical[1,1] - logical variable
%			which equals 1 in case of discrete system
%			and equals 0 in case of continious system

See a difference? Please apply these changes to both help headers and you can merge

**Comment by Vault-Guy** [comment]
To be honest, I see no difference. I have made cosmetic changes in headers but I'm not sure it is something you want to see.

**Comment by pgagarinov** [comment]
Ok, I've uploaded a screenshot (see my comment above) - arrows correspond to tabs, please mind both tabs and spaces.

**Comment by Vault-Guy** [comment]
As I see I've pushed headers with extra tabs in headers - sorry for that -  but I've deleted them in my local branch and have pushed with them into my remote branch. And I've merged from updated origin/remote as had been recommended. And after fetch from origin/master into my local master there are no this annoying tabs.

**Comment by pgagarinov** [comment]
Looks like the task is not done - sym is still used in 
elltool.reach.ReachContinuous class and +elltool\+reach\+test\+mlunit\DiscreteReachTestCase.m

Also varreplace function should be slightly re-designed so that it can an arbitrary variable with an arbitrary variable, not just 't' with 10-t. I.e. varreplace should take 3 inputs: 

varreplace(mCMat,fromVarName,toVarName)

toVarName='10-t' should be formed outside of varreplace function i.e. the first line of varreplace
